CacheMetadata can throw NullPointerException due to concurrency issue.

```shell
...
Caused by: java.lang.NullPointerException: at index 4
    at com.google.common.collect.ObjectArrays.checkElementNotNull (ObjectArrays.java:235)
    at com.google.common.collect.ImmutableSet.construct (ImmutableSet.java:176)
    at com.google.common.collect.ImmutableSet.construct (ImmutableSet.java:201)
    at com.google.common.collect.ImmutableSet.access$000 (ImmutableSet.java:52)
    at com.google.common.collect.ImmutableSet$Builder.build (ImmutableSet.java:637)
    at com.google.cloud.tools.jib.image.ImageLayers$Builder.build (ImageLayers.java:62)
    at com.google.cloud.tools.jib.cache.CacheMetadata.filterLayers (CacheMetadata.java:95)
    at com.google.cloud.tools.jib.cache.CacheReader.getUpToDateLayerBySourceFiles (CacheReader.java:114)
    at com.google.cloud.tools.jib.builder.steps.BuildAndCacheApplicationLayerStep.call (BuildAndCacheApplicationLayerStep.java:115)
...
```

This is because `CacheMetadata#addLayer` is only synchronized with itself, but `CacheMetadata#getLayers` and `CacheMetadata#filterLayers` should be locked with respect to `addLayer` as well.

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheMetadata.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheMetadata.java
@@ -83,7 +83,9 @@ class CacheMetadata {
     }
   }
 
-  ImageLayers<CachedLayerWithMetadata> getLayers() {
+  // TODO: Remove explicit synchronization by refactoring build steps to not mutate a common
+  // CacheMetadata.
+  synchronized ImageLayers<CachedLayerWithMetadata> getLayers() {
     return layersBuilder.build();
   }
 
@@ -91,7 +93,7 @@ class CacheMetadata {
     layersBuilder.add(layer);
   }
 
-  LayerFilter filterLayers() {
+  synchronized LayerFilter filterLayers() {
     return new LayerFilter(layersBuilder.build());
   }
 }

